jenkins-dsl/jenkins_jira_integration.py

```diff
@@ -0,0 +1,352 @@
+            # We know we wrote it; now we need to determine if this comment was for the branch we're currently
+            # processing to determine if we need to update it
+            comment_text = jira.comment('CASSANDRA-' + ticket_number, comment)
+            if 'Branch: ' + branch in comment_text.body:
+                # This is an update
+                if post_to_jira is True:
+                    comment.update(body=ci_results)
+                else:
+                    print('[UPDATE] comment to manually post to Jira for CASSANDRA-' + ticket_number)
+                    print(ci_results)
+                return
+
+    if post_to_jira is True:
+        issue.add_comment('CASSANDRA-' + ticket_number, ci_results)
+    else:
+        print('[ADD] the following comment to Jira for CASSANDRA-' + ticket_number)
+        print(ci_results)
+
+
+def build_local_cache(server: jenkins, branch: str, buildnum: str) -> Dict:
+    """
+    For the input build back as far as we have history, we want to cache the following (k/v):
+        NUMBER:             str number of this build
+        PREVIOUS_NUMBER     str build number of the previous build to walk to when walking the cache
+        failures:           set of test failures seen for this build (note: sets aren't serializable in JSON so we cast to list)
+
+        We use the build number and previous number as strings since they're going to get case to that in the JSON
+        ser/deser anyway, so best not to mix.
+    :return: testnu->
+    """
+    if not os.path.isdir('cache'):
+        os.mkdir('cache')
+
+    cached_data = {}
+    json_file = 'cache/' + branch
+    if os.path.exists(json_file):
+        with open(json_file, 'r', encoding='utf-8') as infile:
+            cached_data = json.load(infile)
+        for build in cached_data:
+            log('Loaded build: [' + str(build) + '] from cache')
+
+    while buildnum != NO_PREVIOUS_BUILD:
+        log('Processing cache for build_number ' + buildnum + ' on branch [' + branch + ']')
+        if buildnum in cached_data:
+            build_data = cached_data[buildnum]
+            log('   Already found data for ' + buildnum + ' cached locally. Continuing to ' + str(build_data[PREVIOUS_NUMBER]))
+            buildnum = str(build_data[PREVIOUS_NUMBER])
+        else:
+            print('   Did not find build_number: ' + buildnum + ' in the json cache. Populating...')
+            # Pull the data from the Jenkins server as we don't have it in our cache
+            build_data = BuildData(server, branch, buildnum)
+
+            # The data in our JSON cache is a pretty simple subset of what we have in BuildData for a full CI report
+            newcache_data = {}
+            newcache_data[NUMBER] = build_data.number
+            newcache_data[PREVIOUS_NUMBER] = build_data.previous_number
+            # Can't serialize sets to JSON
+            newcache_data['failures'] = list(build_data.test_failures)
```

Comment:
does this not include uncompleted builds?  Do we wish to cache non-completed builds?

Reply:
This includes both failed builds and incomplete builds. Since we're not doing a N vs. N-1 comparison and instead comparing against the entire history, the downside to this is our N of M test failure calculation will also include failed builds in it.

Are they frequent enough we expect this to heavily throw off the visual ability to, at a glance, get an intuition of whether a test is a recent failure (1 of M) or intermittent (N of M)?